File: app/src/main/java/com/stupidtree/hichat/ui/group/pick/PickGroupDialog.java
package com.stupidtree.hichat.ui.group.pick;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.stupidtree.hichat.R;
import com.stupidtree.hichat.data.model.RelationGroup;
import com.stupidtree.hichat.ui.base.BaseViewHolder;
import com.stupidtree.hichat.ui.base.BasicSelectableListAdapter;
import com.stupidtree.hichat.ui.base.DataState;
import com.stupidtree.hichat.ui.widgets.TransparentBottomSheetDialog;

import java.util.LinkedList;
import java.util.List;
import java.util.Objects;

import butterknife.BindView;

/**
 * 选择好友分组的底部弹窗
 */
public class PickGroupDialog extends TransparentBottomSheetDialog {

    @BindView(R.id.list)
    RecyclerView list;

    @BindView(R.id.confirm)
    View confirm;

    @BindView(R.id.cancel)
    View cancel;

    @BindView(R.id.loading)
    ProgressBar loading;

    PickGroupViewModel viewModel;
    LAdapter listAdapter;
    OnConfirmListener onConfirmListener;
    String initGroupId = null;
    public interface OnConfirmListener{
        void OnConfirmed(@Nullable RelationGroup group);
    }


    public PickGroupDialog setOnConfirmListener(OnConfirmListener onConfirmListener) {
        this.onConfirmListener = onConfirmListener;
        return this;
    }

    public PickGroupDialog setInitGroupId(String initGroupId) {
        this.initGroupId = initGroupId;
        return this;
    }

    @Override
    protected int getLayoutId() {
        return R.layout.dialog_bottom_group_list;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        viewModel = new ViewModelProvider(this, ViewModelProvider.AndroidViewModelFactory.getInstance(requireActivity().getApplication())).get(PickGroupViewModel.class);
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    protected void initViews(View v) {
        listAdapter = new LAdapter(requireContext(),new LinkedList<>());
        list.setAdapter(listAdapter);
        list.setLayoutManager(new LinearLayoutManager(requireContext()));
        viewModel.getListData().observe(this, listDataState -> {
            loading.setVisibility(View.INVISIBLE);
            if(listDataState.getState()== DataState.STATE.SUCCESS){
                listAdapter.setSelected(listDataState.getData(),initGroupId);
                listAdapter.notifyItemChangedSmooth(listDataState.getData());
            }
        });

        confirm.setOnClickListener(view -> {
            if(onConfirmListener!=null){
                onConfirmListener.OnConfirmed(listAdapter.getSelectedData());
                dismiss();
            }
        });

        cancel.setOnClickListener(view -> dismiss());
    }

    @Override
    public void onResume() {
        super.onResume();
        loading.setVisibility(View.VISIBLE);
        viewModel.startRefresh();
    }

    static class LAdapter extends BasicSelectableListAdapter<RelationGroup, LAdapter.LHolder> {


        public LAdapter(Context mContext, List<RelationGroup> mBeans) {
            super(mContext, mBeans);
        }

        @Override
        protected int getLayoutId(int viewType) {
            return R.layout.dialog_bottom_selectable_list_item;
        }

        @Override
        public LHolder createViewHolder(View v, int viewType) {
            return new LHolder(v);
        }

        @Override
        protected void bindHolder(@NonNull LHolder holder, @Nullable RelationGroup data, int position) {
            if (data != null) {
                holder.text.setText(data.getGroupName());
            }
            if(position==getSelectedIndex()){ //若被选中
                holder.selected.setVisibility(View.VISIBLE);
            }else{
                holder.selected.setVisibility(View.GONE);
            }
            holder.item.setOnClickListener(view -> selectItem(position,data));
        }


        public void setSelected(List<RelationGroup> data,String groupId) {
            mBeans.clear();
            mBeans.addAll(data);
            if(groupId==null&&data.size()>0){
                selectedIndex = 0;
                selectedData = data.get(0);
            }else{
                for(RelationGroup relationGroup:mBeans){
                    if(Objects.equals(relationGroup.getId(),groupId)){
                        selectedIndex = mBeans.indexOf(relationGroup);
                        selectedData = relationGroup;
                    }
                }
            }

        }



        static class LHolder extends BaseViewHolder {
            @BindView(R.id.text)
            TextView text;
            @BindView(R.id.item)
            ViewGroup item;

            @BindView(R.id.selected)
            ImageView selected;
            public LHolder(@NonNull View itemView) {
                super(itemView);
            }
        }

    }

}

File: app/src/main/java/com/stupidtree/cloudliter/ui/group/pick/PickGroupDialog.kt
package com.stupidtree.cloudliter.ui.group.pick

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.viewbinding.ViewBinding
import com.stupidtree.cloudliter.R
import com.stupidtree.cloudliter.data.model.RelationGroup
import com.stupidtree.cloudliter.databinding.DialogBottomGroupListBinding
import com.stupidtree.style.base.BaseViewHolder
import com.stupidtree.component.data.DataState
import com.stupidtree.style.base.BasicSelectableListAdapter
import com.stupidtree.style.databinding.DialogBottomSelectableListItemBinding
import com.stupidtree.style.widgets.TransparentBottomSheetDialog
import java.util.*

/**
 * 选择好友分组的底部弹窗
 */
class PickGroupDialog : TransparentBottomSheetDialog<DialogBottomGroupListBinding>() {
    var viewModel: PickGroupViewModel? = null
    var listAdapter: LAdapter? = null
    private var onConfirmListener: OnConfirmListener? = null
    private var initGroupId: String? = null

    interface OnConfirmListener {
        fun onConfirmed(group: RelationGroup?)
    }

    fun setOnConfirmListener(onConfirmListener: OnConfirmListener?): PickGroupDialog {
        this.onConfirmListener = onConfirmListener
        return this
    }

    fun setInitGroupId(initGroupId: String?): PickGroupDialog {
        this.initGroupId = initGroupId
        return this
    }

    override fun getLayoutId(): Int {
        return R.layout.dialog_bottom_group_list
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        viewModel = ViewModelProvider(this, ViewModelProvider.AndroidViewModelFactory.getInstance(requireActivity().application)).get(PickGroupViewModel::class.java)
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun initViews(v: View) {
        listAdapter = LAdapter(requireContext(), LinkedList())
        binding.list.adapter = listAdapter
        binding.list.layoutManager = LinearLayoutManager(requireContext())
        viewModel!!.listData?.observe(this, { listDataState: DataState<List<RelationGroup>?> ->
            binding.loading.visibility = View.INVISIBLE
            if (listDataState.state === DataState.STATE.SUCCESS) {
                listAdapter!!.setSelected(listDataState.data!!, initGroupId)
                listAdapter!!.notifyItemChangedSmooth(listDataState.data!!)
            }
        })
        binding.confirm.setOnClickListener {
            if (onConfirmListener != null) {
                onConfirmListener!!.onConfirmed(listAdapter!!.selectedData)
                dismiss()
            }
        }
        binding.cancel.setOnClickListener { dismiss() }
    }

    override fun onStart() {
        super.onStart()
        binding.loading.visibility = View.VISIBLE
        viewModel!!.startRefresh()
    }

    class LAdapter(mContext: Context, mBeans: MutableList<RelationGroup>) : BasicSelectableListAdapter<RelationGroup, LAdapter.LHolder>(mContext, mBeans) {

        override fun bindHolder(holder: LHolder, data: RelationGroup?, position: Int) {
            if (data != null) {
                holder.binding.text.text = data.groupName
            }
            if (position == selectedIndex) { //若被选中
                holder.binding.selected.visibility = View.VISIBLE
            } else {
                holder.binding.selected.visibility = View.GONE
            }
            holder.binding.item.setOnClickListener { data?.let { selectItem(position, it) } }
        }

        fun setSelected(data: List<RelationGroup>, groupId: String?) {
            mBeans.clear()
            mBeans.addAll(data)
            if (groupId == null && data.isNotEmpty()) {
                selectedIndex = 0
                selectedData = data[0]
            } else {
                for (relationGroup in mBeans) {
                    if (relationGroup.id == groupId) {
                        selectedIndex = mBeans.indexOf(relationGroup)
                        selectedData = relationGroup
                    }
                }
            }
        }

        class LHolder(itemView: DialogBottomSelectableListItemBinding) : BaseViewHolder<DialogBottomSelectableListItemBinding>(itemView)

        override fun getViewBinding(parent: ViewGroup, viewType: Int): ViewBinding {
            return DialogBottomSelectableListItemBinding.inflate(mInflater, parent, false)
        }

        override fun createViewHolder(viewBinding: ViewBinding, viewType: Int): LHolder {
            return LHolder(viewBinding as DialogBottomSelectableListItemBinding)
        }
    }

    override fun initViewBinding(v: View): DialogBottomGroupListBinding {
        return DialogBottomGroupListBinding.bind(v)
    }
}